### PR TITLE
Create public configuration factory methods

### PR DESCRIPTION
**Pull request by jvandort**
Implements new public lazy factory methods for creating configurations. Additionally created similar internal eager factory methods. We will migrate to the public methods in 9.0

The significance of the new methods is that configuration roles are specified upon construction. This is important for reducing Configuration memory usage, and improving user experience given the large API of the Configuration type.

Fixes: https://github.com/gradle/gradle/issues/9834
Supersedes https://github.com/gradle/gradle/pull/25193

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test qfl

**Comment by jvandort** [comment]
@bot-gradle test apt

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test rfr

**Review by big-guy — APPROVED** [review]

File: subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_dependencies.adoc [hunk 1]
```diff
@@ -82,7 +82,7 @@ include::sample[dir="snippets/dependencyManagement/attributeMatching/groovy",fil
 
 At this point, we have 3 different configurations with different roles:
 
-- `someConfiguration` declares the dependencies of my application. It's just a bucket that can hold a list of dependencies.
+- `someConfiguration` declares the dependencies of my application. It's just a collection of dependencies.
```
Comment (lkasso) [inline subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_dependencies.adoc:85]:
```suggestion
- `someConfiguration` declares the dependencies of my application. It is simply a collection of dependencies.
```

**Review by lkasso — APPROVED** [review]

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Pull request merged**